ipl.1
.TH IPL 1 "06 August 2023" "1.0" "Image Processing Library for Web"
.SH NAME
ipl \- Simple Image Processing Library for Web
.SH SYNOPSIS
.B ipl
.I
.SH DESCRIPTION
.B ipl
.\" ipl is a command-line tool that provides basic image processing functionalities for preparing images
.\" to be used in fast-rendered websites. The user can load an image, resize, crop, rotate, and save the processed
.\" image in various formats. Are you tired of using bloat software just to do simple task an resize, crop or rotate
.\" image, this program is for you!

ipl is a simple command-line tool for image processing, primarily aimed at preparing
images for web usage. It provides various options to resize, crop, and rotate images.
The program supports popular image formats, including PNG, JPEG, BMP, and GIF.
.SH USAGE
.B ipl
ipl is used interactively and will prompt the user for inputs. 
When prompted to enter the image's name, you can use an absolute path, such as 
/path/to/image.png, or a relative path from the directory where you call the 
ipl program. For example, if you are in the root directory of your project and 
want to process an image located at assets/img/logo.png, you can simply enter 
assets/img/logo.png as the image name. Similarly, you can enter relative paths 
using the ../ notation. For example, if the image is located two directories 
above the current directory, you can enter ../../directory/image.png.
.SH CROP
.TP
.BR -c
Crop a specific region from an image. This option allows you to select a rectangular
region within the image and create a new image containing only that region. When prompted,
you can specify the X and Y coordinates of the top-left corner of the region, as well as
the width and height of the region.
.TP
.BR -x
The X coordinate of the top-left corner for cropping or resizing.

.TP
.BR -y 
The Y coordinate of the top-left corner for cropping or resizing.

.TP
.BR -b 
Use with cropping to specify an equal border for cropping from all sides.
Just specify border without 'x' and 'y'.

.SH RESIZE
.TP
.BR -r 
Resize an image to a custom size.

.TP
.BR --background
Resize an image for use as a background. This option resizes the image to fit the screen,
considering common aspect ratios for desktop backgrounds.

.TP
.BR --hero
Resize an image to be used as a hero image on a website. This option resizes the image to
common hero image dimensions, suitable for banners and headers.

.TP
.BR --web-banner
Resize an image for use as a website banner. The recommended dimensions are 250 x 250 with
an aspect ratio of 1:1.

.TP
.BR --blog
Resize an image for use in a blog post. The recommended dimensions are 1200 x 630 with
an aspect ratio of 3:2.

.TP
.BR --logo-rec
Resize an image for use as a rectangular logo. The recommended dimensions are 250 x 100 with
an aspect ratio of 2:3.

.TP
.BR --logo-sq
Resize an image for use as a square logo. The recommended dimensions are 100 x 100 with
an aspect ratio of 1:1.

.TP
.BR --favicon
Resize an image for use as a favicon. The recommended dimensions are 16 x 16 with an
aspect ratio of 1:1.

.TP
.BR --social
Resize an image for use as a social media icon. The recommended dimensions are 32 x 32 with
an aspect ratio of 1:1.

.TP
.BR --lightbox
Resize an image for use in a lightbox (full-screen view). The recommended dimensions are
1600 x 500 with an aspect ratio of 16:9.

.TP
.BR --thumb
Resize an image for use as a thumbnail. The recommended dimensions are 150 x 150 with
an aspect ratio of 1:1.

.TP
.BR --custom
Resize an image to a custom size. Use this option if you want to specify your own width and
height for resizing.

.TP
.BR -w 
Specify the width for resizing.

.TP
.BR -h 
Specify the height for resizing.
.SH EXAMPLES
Resize an image to a custom size:
.nf
.BR $ ipl
Enter path to picture: /path/to/input/image.png
Choose option:
1 Resize image
2 Crop image
3 Rotate image
4 Quit without saving
Option number: 1
Resize option number: 11
Enter WIDTH [pixels]: 800
Enter HEIGHT [pixels]: 600
Enter new name [example.png]: /path/to/output/resized_image.png
Proceed with new dimensions? [Y,n] Y
New photo size is 1440000 B
Successfully changed dimensions! Image saved to /path/to/output/resized_image.png!
.SH AUTHORS
.B ipl
was developed by Vidoje Šević <vidoje.sevic.275.19@ict.edu.rs>.
.SH SEE ALSO
.BR stb_resize(3),
.BR stb_image(3),
.BR stb_image_write(3)
.SH BUGS
Report any bugs or issues to <vidoje.sevic.275.19@ict.edu.rs>.



ipl.c
/*
 * ipl - Simple Image Processing Library for web
 * Copyright (C) 2023 Vidoje Šević
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/gpl-3.0.txt>
 * */

// UNIX PORN

#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <stdbool.h>
#include <ctype.h>
#include <getopt.h>
#include "ipl.h"
#include "cli.h"
#include "prompt.h"
#include "resize.h"
#include "crop.h"
#include "rotate.h"

//Dependencies
#define STB_IMAGE_IMPLEMENTATION
#include "stb/stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb/stb_image_write.h"

int main(int argc, char *argv[]) {
    Input input;
    Pics img;
    Dime dime;
    unsigned char *imgData;
    char *fileSize = NULL;
    Crop crop;

    if (argc > 1) {
        // If user call ./ipl with more than 1 argv
        parseArguments(argc, argv, &img, &input, &crop);
    } else {
        // prompt mode
        int option = 0;
        printf("Welcome to simple Image Processing Library for web!\n");
        printf("In few quick steps your image will be ready for fast rendered website!\n");

        printf("Enter path to picture: ");

        // Checking if path is correct
        checkName(&img);
        getPath(&img);

        allocateImg(img.path, &input, &imgData);

        // Calculate image size in KiB
        getSize(img.path, &input, &fileSize);

        // Copy the loaded image data to the allocated memory
        img.data = input.data;
        img.width = input.width;
        img.height = input.height;
        img.channel = input.channels;
        img.bitDepth = input.bitDepth;
        strcpy(img.bitInfo, input.bitInfo);
        img.size = fileSize;

        printInfo(&img);
        
        promptMode(&img, &dime, option, &crop);

        // stbi_image_free(imgData);
        free(fileSize);
        free(imgData);
        free(input.data);
        free(img.data);
        free(img.size);
    }


    return 0;
}

Pics* getPath(Pics *img) {

    if (strlen(img->path) == 0) {
        fprintf(stderr, "No input provided. Please enter a valid image path.\n");
        exit(EXIT_FAILURE);
    }
    
    findOutExtension(img->path, img->ext);

    // Get extension

    return img;
}

void findOutExtension(char *path, char extension[EXT_SIZE]){
    extension[0] = '\0';
    const char *supportedExtensions[] = {"jpg", "jpeg", "png", "bmp", "tga", "hdr"};
    int isValidExtension = 0;

    char *ext = strrchr(path, '.');
    if (ext) {
        ext++;
        for (int i = 0; i < sizeof(supportedExtensions) / sizeof(supportedExtensions[0]); ++i) {
            if (strcmp(ext, supportedExtensions[i]) == 0) {
                strncpy(extension, supportedExtensions[i], EXT_SIZE);
                break;
            }
        }
    }
    if (!ext) {
        isValidExtension = 1;
    }

    if (isValidExtension != 0) {
        fprintf(stderr, "Unsupported file extension. Supported formats: .jpg, .jpeg, .png, .bmp, .tga, .hdr!\n");
        exit(EXIT_FAILURE);
    }
}

void allocateImg(char *path, Input *input, unsigned char **imgData) {
    *imgData = stbi_load(path, &input->width, &input->height, &input->channels, 0);

    if (*imgData == NULL) {
        fprintf(stderr, "Error loading image: %s from '%s'!\n", stbi_failure_reason(), path);
        stbi_image_free(imgData); // Free the allocated memory before exiting
        exit(EXIT_FAILURE);
    } else {
        input->data = (unsigned char*)malloc(input->width * input->height * input->channels);
    }

    if (input->data == NULL) {
        fprintf(stderr, "Error: Memory allocation failed!\n");
        stbi_image_free(*imgData);
        exit(EXIT_FAILURE);
    }

    memcpy(input->data, *imgData, input->width * input->height * input->channels);

    free(*imgData);
}

void getSize(char *path, Input *input, char **fileSize) {
    // Detect bit per color
    int bpp = stbi_info(path, &input->width, &input->height, &input->bitDepth);

    if (input->channels == 1) {
        input->bitDepth = 1;
    } else if (input->channels == 3) {
        input->bitDepth = 8;
    } else if (input->channels == 4) {
        input->bitDepth = 32;
    } else {
        // For other channel counts, we'll check the actual bit depth using stbi_info
        if (bpp == 8) {
            input->bitDepth = 8;
        }
        else if (bpp == 16) {
            input->bitDepth = 16;
        }
        else if (bpp == 24) {
            input->bitDepth = 24;
        }
        else {
            input->bitDepth = 0;
        }
    }

    if (input->bitDepth == 1) {
    }
    if (input->bitDepth == 8) {
        strcpy(input->bitInfo, "8-bit/color (RGB or grayscale, 3 channels)");
    }
    if (input->bitDepth == 16) {
        strcpy(input->bitInfo, "16-bit/color (RGB, 3 channels)");
    }
    if (input->bitDepth == 24) {
        strcpy(input->bitInfo, "24-bit/color (RGB, 3 channels)");
    }
    if (input->bitDepth == 32) {
        strcpy(input->bitInfo, "32-bit/color (RGBA, 4 channels)");
    }

    // if (*fileSize != NULL) {
    //     free(*fileSize);
    // }

    *fileSize = calcSize(path);
}

char* calcSize(const char* filename) {
    long fileSize;
    char sizeChar[SIZE];
    char *size = NULL;
    struct stat st;

    if (stat(filename, &st) == 0) {
        fileSize = st.st_size;
    } else {
        fileSize = 0;
    }

    if (fileSize > KiloByte) {
        fileSize = fileSize / KiloByte;
        snprintf(sizeChar, sizeof(sizeChar), "%ld KiB", fileSize);
    }
    else {
        snprintf(sizeChar, sizeof(sizeChar), "%ld B", fileSize);
    }

    size = strdup(sizeChar);
    if (size == NULL) {
        perror("Error alocation failed");
        exit(EXIT_FAILURE);
        free(size);
    }

    return size;
}

void getWidth(Dime *dime) {
    char input[BIT_NAME_SIZE];
    int validInput = 0;

    while (!validInput) {
        if (fgets(input, sizeof(input), stdin) == NULL) {
            perror("Error reading input.");
            exit(EXIT_FAILURE);
        }
        if (sscanf(input, "%d", &dime->resWidth) != 1) {
            printf("Invalid input. Please enter a valid integer.\n");
            continue;
        }
        validInput = 1;
    }
}

void getHeight(Dime *dime) {
    char input[BIT_NAME_SIZE];
    int validInput = 0;

    while (!validInput) {
        if (fgets(input, sizeof(input), stdin) == NULL) {
            perror("Error reading input.");
            exit(EXIT_FAILURE);
        }
        if (sscanf(input, "%d", &dime->resHeight) != 1) {
            printf("Invalid input. Please enter a valid integer.\n");
            continue;
        }
        validInput = 1;
    }
}

void getName(Dime *dime) {
    char input[BIT_NAME_SIZE];
    int validInput = 0;

    while (!validInput) {
        printf("Enter new name [example.png]: ");
        if (fgets(input, sizeof(input), stdin) == NULL) {
            perror("Error reading input.");
            exit(EXIT_FAILURE);
        }
        input[strcspn(input, "\n")] = '\0';
        if (strlen(input) == 0) {
            printf("Invalid input. Please enter a valid string.\n");
            validInput = 1;
        } else {
            strncpy(dime->name, input, sizeof(dime->name) - 1);
            validInput = 1;
        }
    }
    findOutExtension(dime->name, dime->ext);
    // printf("Dime name is %s, ext is %s\n", dime->name, dime->ext);
}

void saveImage(unsigned char* imageData, int width, int height, int channel, const char* filename, char *extension, int quality) {

    int result = 0;
    if (strcmp(extension, "jpg") == 0 || strcmp(extension, "jpeg") == 0) {
        saveJPG(imageData, width, height, channel, filename, &result, quality);
    } else if (strcmp(extension, "png") == 0) {
        savePNG(imageData, width, height, channel, filename, &result);
    } else if (strcmp(extension, "bmp") == 0) {
        saveBMP(imageData, width, height, channel, filename, &result);
    } else if (strcmp(extension, "tga") == 0) {
        saveTGA(imageData, width, height, channel, filename, &result);
    } else if (strcmp(extension, "hdr") == 0) {
        saveHDR(imageData, width, height, channel, filename, &result);
    } else {
        printf("Unsupported file extension.\n");
        free(imageData);
        return;
    }

    char *sizeChar;

    if (result == 0) {
        printf("Error saving image to %s.\n", filename);
    } else {
        sizeChar = calcSize(filename);
        printf("Image saved to %s! Size: %s!\n", filename, sizeChar);
        free(sizeChar);
    }
    free(imageData);
}

void saveJPG(unsigned char* imageData, int width, int height, int channel, const char* filename, int *result, int quality) {
    // if (quality == 0) {
    //     quality = 80;
    // }
    if (quality > 0 && quality <= 100)
        *result = stbi_write_jpg(filename, width, height, channel, imageData, quality);
}

void savePNG(unsigned char* imageData, int width, int height, int channel, const char* filename, int *result) {
    *result = stbi_write_png(filename, width, height, channel, imageData, width * channel);
}

void saveBMP(unsigned char* imageData, int width, int height, int channel, const char* filename, int *result) {
    *result = stbi_write_bmp(filename, width, height, channel, imageData);
}

void saveTGA(unsigned char* imageData, int width, int height, int channel, const char* filename, int *result) {
    *result = stbi_write_tga(filename, width, height, channel, imageData);
}

void saveHDR(unsigned char* imageData, int width, int height, int channel, const char* filename, int *result) {
    float *floatImageData = (float *)malloc(width * height * channel * sizeof(float));

    // Convert unsigned char image data to float
    for (int i = 0; i < width * height * channel; ++i) {
        floatImageData[i] = (float)imageData[i] / 255.0f;  // Assuming imageData is in range 0-255
    }

    *result = stbi_write_hdr(filename, width, height, channel, floatImageData);

    free(floatImageData);
}

void clearInputBuffer() {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}

